chain/indexer/src/streamer/streamer.rs

```diff
@@ -0,0 +1,228 @@
+async fn fetch_outcomes(
+    client: &Addr<near_client::ViewClientActor>,
+    mut outcomes_drain: impl Iterator<Item = types::TransactionOrReceiptId>,
+) -> (Vec<Outcome>, Vec<types::TransactionOrReceiptId>) {
+    let mut outcomes: Vec<Outcome> = vec![];
+    let mut outcomes_to_retry: Vec<types::TransactionOrReceiptId> = vec![];
+    while let Some(transaction_or_receipt_id) = outcomes_drain.next() {
+        match client
+            .send(near_client::GetExecutionOutcome { id: transaction_or_receipt_id.clone() })
+            .await
+            .map_err(|_| FailedToFetchData)
+        {
+            Ok(Ok(outcome)) => match &transaction_or_receipt_id {
+                types::TransactionOrReceiptId::Transaction { .. } => {
+                    outcomes.push(Outcome::Transaction(outcome.outcome_proof));
+                }
+                types::TransactionOrReceiptId::Receipt { .. } => {
+                    outcomes.push(Outcome::Receipt(outcome.outcome_proof));
+                }
+            },
+            _ => {
+                outcomes_to_retry.push(transaction_or_receipt_id);
+            }
+        }
+    }
+    (outcomes, outcomes_to_retry)
+}
+
+/// Fetches all the chunks by their hashes.
+/// Includes transactions and receipts in custom struct (to provide more info).
+/// Returns Chunks as a `Vec`
+async fn fetch_chunks(
+    client: &Addr<near_client::ViewClientActor>,
+    chunks: &[views::ChunkHeaderView],
+) -> Result<Vec<views::ChunkView>, FailedToFetchData> {
+    let chunks_hashes =
+        chunks.iter().map(|chunk| near_client::GetChunk::ChunkHash(chunk.chunk_hash.into()));
+    let mut chunks: futures::stream::FuturesUnordered<_> =
+        chunks_hashes.map(|get_chunk| fetch_single_chunk(&client, get_chunk)).collect();
+    let mut response = Vec::<views::ChunkView>::with_capacity(chunks.len());
+
+    while let Some(chunk) = chunks.next().await {
+        response.push(chunk?);
+    }
+
+    Ok(response)
+}
+
+/// Function that starts Streamer's busy loop. Every half a seconds it fetches the status
+/// compares to already fetched block height and in case it differs fetches new block of given height.
+///
+/// We have to pass `client: Addr<near_client::ClientActor>` and `view_client: Addr<near_client::ViewClientActor>`.
+pub(crate) async fn start(
+    view_client: Addr<near_client::ViewClientActor>,
+    client: Addr<near_client::ClientActor>,
+    mut blocks_sink: mpsc::Sender<BlockResponse>,
+) {
+    info!(target: INDEXER, "Starting Streamer...");
+    let mut outcomes_to_get = Vec::<types::TransactionOrReceiptId>::new();
+    let mut last_synced_block_height: types::BlockHeight = 0;
+    'main: loop {
+        time::delay_for(INTERVAL).await;
+        let status = fetch_status(&client).await;
+        if let Ok(status) = status {
+            if status.sync_info.syncing {
+                continue;
+            }
+        }
+
+        let block = if let Ok(block) = fetch_latest_block(&view_client).await {
```

Comment:
Let's fetch the final (`Finality::Final`) block instead of latest to avoid dealing with forks until we know how we want to handle the forks.

Reply:
https://github.com/nearprotocol/nearcore/pull/2651/files#diff-2077d50fa94c2f282718b999f1945dafR62

I don't think we should change the function name, because it is latest more-or-less safe block for Indexer to fetch.

Reply:
Sounds good to me.